Src/ILGPU.Algorithms.Tests/ArrayExtensionTests.tt

```diff
@@ -0,0 +1,145 @@
+
+namespace ILGPU.Algorithms.Tests
+{
+
+    public abstract partial class ArrayExtensionTests : TestBase
+    {
+        protected ArrayExtensionTests(ITestOutputHelper output, TestContext testContext)
+            : base(output, testContext)
+        { }
+
+        #region MemberData
+
+        public static TheoryData<object, object> TestDataLength1D =>
+            new TheoryData<object, object>
+        {
+<#
+        int value = 0; 
+        foreach (var type in NumericTypes) {
+#>
+            { (<#= type.Type #>)<#= value #>, 31 },
+            { (<#= type.Type #>)<#= value #>, 63 },
+<#
+            value++;
+        }
+#>
+        };
+
+        public static TheoryData<object, object, object> TestDataLength2D =>
+            new TheoryData<object, object, object>
+        {
+<#
+        foreach (var type in NumericTypes) {
+#>
+            { (<#= type.Type #>)<#= value #>, 31, 63 },
+            { (<#= type.Type #>)<#= value #>, 32, 64 },
+<#
+            value++;
+        }
+#>
+        };
+
+        public static TheoryData<object, object, object, object> TestDataLength3D =>
+            new TheoryData<object, object, object, object>
+        {
+<#
+        foreach (var type in NumericTypes) {
+#>
+            { (<#= type.Type #>)<#= value #>, 31, 63, 127 },
+            { (<#= type.Type #>)<#= value #>, 32, 64, 128 },
+<#
+            value++;
+        }
+#>
+        };
+
+        #endregion
+
+        [Theory]
+        [MemberData(nameof(TestDataLength1D))]
+        public void ArrayGetExtent1D<T>(T _, int lenX)
+            where T : unmanaged
+        {
+            T[] dummy = new T[lenX];
+            Index1 actual = ArrayExtensions.GetExtent(dummy);
+
+            Assert.Equal(lenX, actual.X);
+        }
+
+        [Theory]
+        [MemberData(nameof(TestDataLength2D))]
+        public void ArrayGetExtent2D<T>(T _, int lenX, int lenY)
+            where T : unmanaged
+        {
+            T[,] dummy = new T[lenX, lenY];
+            Index2 actual = ArrayExtensions.GetExtent(dummy);
+
+            Assert.Equal(lenX, actual.X);
+            Assert.Equal(lenY, actual.Y);
+        }
+
+        [Theory]
+        [MemberData(nameof(TestDataLength3D))]
+        public void ArrayGetExtent3D<T>(T _, int lenX, int lenY, int lenZ)
+            where T : unmanaged
+        {
+            T[,,] dummy = new T[lenX, lenY, lenZ];
+            Index3 actual = ArrayExtensions.GetExtent(dummy);
+
+            Assert.Equal(lenX, actual.X);
+            Assert.Equal(lenY, actual.Y);
+            Assert.Equal(lenZ, actual.Z);
+        }
+
+        [Theory]
+        [MemberData(nameof(TestDataLength1D))]
+        public void ArraySetGetValue1D<T>(T value, int lenX)
+            where T : unmanaged
+        {
+            T[] dummy = new T[lenX];
+            Index1 index = lenX/2;
+            ArrayExtensions.SetValue(dummy, value, index);
+            var actual = ArrayExtensions.GetValue(dummy, index);
+            Assert.Equal(value, actual);
```

Comment:
nit: the code above uses empty lines between `actual` and the `Assert...` code. Is this intended?